--- NetworkManager-0.9.1.90/src/nm-system.c.wiget	2011-09-24 18:28:45.765344615 +0200
+++ NetworkManager-0.9.1.90/src/nm-system.c	2011-09-24 18:28:52.295048440 +0200
@@ -40,7 +40,6 @@
 #include <netdb.h>
 #include <glib.h>
 #include <ctype.h>
-#include <net/if.h>
 
 #include "nm-system.h"
 #include "nm-device.h"
--- NetworkManager-0.9.1.90/src/tests/Makefile.am.wiget	2011-09-24 18:35:29.437015456 +0200
+++ NetworkManager-0.9.1.90/src/tests/Makefile.am	2011-09-24 18:36:32.937462055 +0200
@@ -19,6 +19,7 @@ test_dhcp_options_SOURCES = \
 test_dhcp_options_CPPFLAGS = \
 	$(GLIB_CFLAGS) \
 	$(DBUS_CFLAGS) \
+	$(LIBNL_LIBS) \
 	-DDHCLIENT_PATH=\"$(DHCLIENT_PATH)\" \
 	-DDHCPCD_PATH=\"$(DHCPCD_PATH)\"
 
@@ -26,6 +27,7 @@ test_dhcp_options_LDADD = \
 	$(top_builddir)/libnm-util/libnm-util.la \
 	$(top_builddir)/src/dhcp-manager/libdhcp-manager.la \
 	$(top_builddir)/src/libtest-dhcp.la \
+	$(LIBNL_LIBS) \
 	$(GLIB_LIBS) \
 	$(DBUS_LIBS)
 
--- NetworkManager-0.9.1.90/src/nm-device.c.wiget	2011-09-24 18:27:57.987511424 +0200
+++ NetworkManager-0.9.1.90/src/nm-device.c	2011-09-24 18:36:58.819618527 +0200
@@ -25,7 +25,6 @@
 #include <dbus/dbus.h>
 #include <netinet/in.h>
 #include <string.h>
-#include <net/if.h>
 #include <unistd.h>
 #include <errno.h>
 #include <sys/ioctl.h>
--- NetworkManager-0.9.1.90/configure.ac.wiget	2011-09-24 18:38:08.216462329 +0200
+++ NetworkManager-0.9.1.90/configure.ac	2011-09-24 18:39:35.251907185 +0200
@@ -338,8 +338,8 @@ if (test "${have_libnl2}" = "yes"); then
 	have_libnl="yes"
 fi
 
-PKG_CHECK_MODULES(LIBNL3, libnl-3.0, [have_libnl2=yes], [have_libnl2=no])
-if (test "${have_libnl2}" = "yes"); then
+PKG_CHECK_MODULES(LIBNL3, libnl-3.0 libnl-route-3.0, [have_libnl3=yes], [have_libnl3=no])
+if (test "${have_libnl3}" = "yes"); then
 	AC_DEFINE(HAVE_LIBNL3, 1, [Define if you require specific libnl-3 support])
 	LIBNL_CFLAGS="$LIBNL3_CFLAGS"
 	LIBNL_LIBS="$LIBNL3_LIBS"
