# there's no <linux/if_ppp.h> in linux-libc-headers-3.4.0
--- NetworkManager-0.9.4.0/src/ppp-manager/nm-ppp-manager.c.orig	2012-06-14 09:42:51.793598842 +0200
+++ NetworkManager-0.9.4.0/src/ppp-manager/nm-ppp-manager.c	2012-06-14 09:46:36.876054350 +0200
@@ -32,14 +32,14 @@
 #include <sys/socket.h>
 #include <sys/ioctl.h>
 #include <asm/types.h>
-#include <linux/if.h>
+#include <net/if.h>
 #include <sys/stat.h>
 
-#include <linux/ppp_defs.h>
+#include <net/ppp_defs.h>
 #ifndef aligned_u64
 #define aligned_u64 unsigned long long __attribute__((aligned(8)))
 #endif
-#include <linux/if_ppp.h>
+#include <net/if_ppp.h>
 
 #include "NetworkManager.h"
 #include "nm-glib-compat.h"
