diff -urN NetworkManager-0.8.2/initscript/Makefile.am NetworkManager-0.8.2.new//initscript/Makefile.am
--- NetworkManager-0.8.2/initscript/Makefile.am	2010-10-19 01:44:05.000000000 +0200
+++ NetworkManager-0.8.2.new//initscript/Makefile.am	2010-11-04 23:09:47.487444761 +0100
@@ -26,3 +26,6 @@
 if TARGET_LINEXA
 SUBDIRS += linexa
 endif
+if TARGET_PLD
+SUBDIRS += PLD
+endif
diff -urN NetworkManager-0.8.2/initscript/PLD/Makefile.am NetworkManager-0.8.2.new//initscript/PLD/Makefile.am
--- NetworkManager-0.8.2/initscript/PLD/Makefile.am	1970-01-01 01:00:00.000000000 +0100
+++ NetworkManager-0.8.2.new//initscript/PLD/Makefile.am	2010-11-04 21:54:27.000000000 +0100
@@ -0,0 +1,5 @@
+EXTRA_DIST = NetworkManager
+DISTCLEANFILES = NetworkManager
+
+initddir = $(sysconfdir)/rc.d/init.d
+initd_SCRIPTS = NetworkManager
diff -urN NetworkManager-0.8.2/initscript/PLD/NetworkManager.in NetworkManager-0.8.2.new//initscript/PLD/NetworkManager.in
--- NetworkManager-0.8.2/initscript/PLD/NetworkManager.in	1970-01-01 01:00:00.000000000 +0100
+++ NetworkManager-0.8.2.new//initscript/PLD/NetworkManager.in	2010-11-04 21:54:27.000000000 +0100
@@ -0,0 +1,75 @@
+#!/bin/sh
+#
+# NetworkManager	NetworkManager daemon
+#
+# chkconfig:		345 99 01
+#
+# description:		This is a daemon for automatically switching network \
+#			connections to the best available connection.
+#
+# processname:		NetworkManager
+# pidfile:		/var/run/NetworkManager.pid
+#
+
+# Source function library.
+. /etc/rc.d/init.d/functions
+
+upstart_controlled
+
+# Source networking configuration.
+. /etc/sysconfig/network
+
+# Check that networking is up.
+if is_yes "${NETWORKING}"; then
+	if [ ! -f /var/lock/subsys/network -a "$1" != stop -a "$1" != status ]; then
+		msg_network_down NetworkManager
+		exit 1
+	fi
+else
+	exit 0
+fi
+
+start() {
+	# Check if the service is already running?
+	if [ ! -f /var/lock/subsys/NetworkManager ]; then
+		msg_starting NetworkManager
+		daemon NetworkManager --pid-file=/var/run/NetworkManager.pid
+		RETVAL=$?
+		[ $RETVAL -eq 0 ] && touch /var/lock/subsys/NetworkManager
+	else
+		msg_already_running NetworkManager
+	fi
+}
+stop() {
+	if [ -f /var/lock/subsys/NetworkManager ]; then
+		# Stop daemons.
+		msg_stopping NetworkManager
+		killproc NetworkManager
+		rm -f /var/lock/subsys/NetworkManager
+	else
+		msg_not_running NetworkManager
+	fi
+}
+
+RETVAL=0
+case "$1" in
+  start)
+	start
+	;;
+  stop)
+	stop
+	;;
+  restart)
+	stop
+	start
+	;;
+  status)
+	status NetworkManager
+	RETVAL=$?
+	;;
+  *)
+	msg_usage "$0 {start|stop|restart|status}"
+	exit 3
+esac
+	        
+exit $RETVAL
diff -urN NetworkManager-0.8.2/src/backends/NetworkManagerPLD.c NetworkManager-0.8.2.new//src/backends/NetworkManagerPLD.c
--- NetworkManager-0.8.2/src/backends/NetworkManagerPLD.c	1970-01-01 01:00:00.000000000 +0100
+++ NetworkManager-0.8.2.new//src/backends/NetworkManagerPLD.c	2010-11-04 21:54:27.000000000 +0100
@@ -0,0 +1,62 @@
+/* NetworkManager -- Network link manager
+ *
+ * Patryk Zawadzki <patrys@pld-linux.org>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ *
+ * (C) Copyright 2004 RedHat, Inc.
+ * (C) Copyright 2004 Narayan Newton
+ * (C) Copyright 2005 wrobell <wrobell@pld-linux.org>
+ * (C) Copyright 2007 Marcin Banasiak <megabajt@pld-linux.org>
+ * (C) Copyright 2005-2008 Patryk Zawadzki <patrys@pld-linux.org>
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "NetworkManagerGeneric.h"
+#include "nm-system.h"
+#include "NetworkManagerUtils.h"
+
+/*
+ * nm_system_enable_loopback
+ *
+ * Bring up the loopback interface
+ *
+ */
+void nm_system_enable_loopback (void)
+{
+	nm_generic_enable_loopback ();
+}
+
+/*
+ * nm_system_update_dns
+ *
+ * Invalidate the nscd host cache, if it exists, since
+ * we changed resolv.conf.
+ *
+ */
+void nm_system_update_dns (void)
+{
+	if (g_file_test ("/usr/sbin/nscd", G_FILE_TEST_EXISTS | G_FILE_TEST_IS_EXECUTABLE | G_FILE_TEST_IS_REGULAR)) {
+		nm_spawn_process ("/usr/sbin/nscd -i hosts");
+	}
+}
+
diff -urN NetworkManager-0.8.996/configure.ac NetworkManager-0.8.996.new//configure.ac
--- NetworkManager-0.8.996/configure.ac	2011-03-10 07:44:53.000000000 +0000
+++ NetworkManager-0.8.996.new//configure.ac	2011-03-11 21:56:12.369999140 +0000
@@ -95,7 +95,7 @@
 dnl
 AC_C_BIGENDIAN
 
-AC_ARG_WITH(distro, AS_HELP_STRING([--with-distro=DISTRO], [Specify the Linux distribution to target: One of redhat, suse, gentoo, debian, arch, slackware, paldo, mandriva, pardus, linexa or exherbo]))
+AC_ARG_WITH(distro, AS_HELP_STRING([--with-distro=DISTRO], [Specify the Linux distribution to target: One of redhat, suse, gentoo, debian, arch, slackware, paldo, mandriva, pardus, linexa, exherbo or pld]))
 if test "z$with_distro" = "z"; then
 	AC_CHECK_FILE(/etc/redhat-release,with_distro="redhat")
 	AC_CHECK_FILE(/etc/SuSE-release,with_distro="suse")
@@ -109,6 +109,7 @@
 	AC_CHECK_FILE(/etc/pardus-release,with_distro="pardus")
 	AC_CHECK_FILE(/etc/linexa-release,with_distro="linexa")
 	AC_CHECK_FILE(/etc/exherbo-release,with_distro="exherbo")
+	AC_CHECK_FILE(/etc/pld-release,with_distro="pld")
 	if test "z$with_distro" = "z"; then
 		with_distro=`lsb_release -is`
 	fi
@@ -120,7 +121,7 @@
 	exit 1
 else
 	case $with_distro in
-		redhat|suse|gentoo|debian|slackware|arch|paldo|frugalware|mandriva|pardus|linexa|exherbo) ;;
+		redhat|suse|gentoo|debian|slackware|arch|paldo|frugalware|mandriva|pardus|linexa|exherbo|pld) ;;
 		*)
 			echo "Your distribution (${with_distro}) is not yet supported!  (patches welcome)"
 			exit 1
@@ -188,6 +189,11 @@
   AC_DEFINE(TARGET_EXHERBO, 1, [Define if you have Exherbo])
 fi
 
+AM_CONDITIONAL(TARGET_PLD, test x"$with_distro" = xpld)
+if test x"$with_distro" = xpld; then
+  AC_DEFINE(TARGET_PLD, 1, [Define if you have PLD])
+fi
+
 dnl
 dnl Distribution version string
 dnl
@@ -617,6 +623,8 @@
 initscript/paldo/NetworkManager
 initscript/Mandriva/Makefile
 initscript/Mandriva/networkmanager
+initscript/PLD/Makefile
+initscript/PLD/NetworkManager
 initscript/linexa/Makefile
 initscript/linexa/networkmanager
 introspection/Makefile
diff -urN NetworkManager-0.8.996/src/backends/Makefile.am NetworkManager-0.8.996.new//src/backends/Makefile.am
--- NetworkManager-0.8.996/src/backends/Makefile.am	2011-01-02 23:18:57.000000000 +0000
+++ NetworkManager-0.8.996.new//src/backends/Makefile.am	2011-03-11 21:57:30.163333077 +0000
@@ -61,6 +61,10 @@
 libnmbackend_la_SOURCES += NetworkManagerExherbo.c
 endif
 
+if TARGET_PLD
+libnmbackend_la_SOURCES += NetworkManagerPLD.c
+endif
+
 libnmbackend_la_LIBADD += \
 	$(top_builddir)/src/logging/libnm-logging.la \
 	$(DBUS_LIBS) \
diff -urN NetworkManager-0.8.996/src/settings/plugins/Makefile.am NetworkManager-0.8.996.new//src/settings/plugins/Makefile.am
--- NetworkManager-0.8.996/src/settings/plugins/Makefile.am	2011-03-08 16:19:01.000000000 +0000
+++ NetworkManager-0.8.996.new//src/settings/plugins/Makefile.am	2011-03-11 21:59:26.183332964 +0000
@@ -4,6 +4,10 @@
 SUBDIRS+=ifcfg-rh
 endif
 
+if TARGET_PLD
+SUBDIRS+=ifcfg-rh
+endif
+
 if TARGET_SUSE
 SUBDIRS+=ifcfg-suse
 endif
