--- NetworkManager-1.10.0/configure.ac.orig	2017-12-09 18:16:02.242310397 +0100
+++ NetworkManager-1.10.0/configure.ac	2017-12-09 18:17:47.650126844 +0100
@@ -1260,7 +1260,7 @@
 build_docs=no
 if test -n "$INTROSPECTION_MAKEFILE"; then
 	# If g-i is installed we know we have python, but we might not have pygobject
-	if ! "$PYTHON" -c 'from gi.repository import GObject' >& /dev/null; then
+	if ! "$PYTHON" -c 'from gi.repository import GObject' > /dev/null 2>&1; then
 		AC_MSG_ERROR(["--enable-introspection aims to build the settings documentation. This requires GObject introspection for python (pygobject)])
 	fi
 
