diff -urN NetworkManager-0.6.5.orig/gnome/vpn-properties/Makefile.am NetworkManager-0.6.5/gnome/vpn-properties/Makefile.am
--- NetworkManager-0.6.5.orig/gnome/vpn-properties/Makefile.am	2007-07-23 19:44:13.000000000 +0200
+++ NetworkManager-0.6.5/gnome/vpn-properties/Makefile.am	2007-09-07 23:36:33.000000000 +0200
@@ -26,7 +26,6 @@
         -DGLADEDIR=\""$(gladedir)"\"                    \
         -DG_DISABLE_DEPRECATED                          \
         -DGDK_DISABLE_DEPRECATED                        \
-        -DGNOME_DISABLE_DEPRECATED                      \
 	-DGNOMELOCALEDIR=\"$(datadir)/locale\"		\
         -DVERSION=\"$(VERSION)\" \
         $(NULL)
